### PR TITLE
Handle ref-containing structs returned by value from function-pointers

### PR DESCRIPTION
**Pull request by 333fred**
My fix for https://github.com/dotnet/roslyn/issues/49315 missed the case of ref-containing types like Span being returned by value, instead of by reference. We now check that case as well. Fixes https://github.com/dotnet/roslyn/issues/49760.

### CODE REVIEW

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for review.

**Comment by AlekseyTs** [comment]
>                 case BoundKind.Call:

Please look for other instances where we perform similar kind check or a type check and confirm that we are not missing a check for function pointers.

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:2041 in b1dc76d. [](commit_id = b1dc76d622d6af3d0e158d9d8a6eb2906ad17c1e, deletion_comment = False)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 1)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs [hunk 1]
```diff
@@ -10999,6 +10999,119 @@ .maxstack  2
...
+using System;
+unsafe
+{
+    static Span<int> ptrTest()
+    {
+        Span<int> s = stackalloc int[1];
+        delegate*<Span<int>, Span<int>> ptr = &test;
+        return ptr(s);
+    }
+
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:11051]: Have the positive case as well to make sure escape is hooked up correctly here.

```suggestion

    static Span<int> ptrTest2(Span<int> s)
    {
        delegate*<Span<int>, Span<int>> ptr = &test;
        return ptr(s);
    }
```

**Review by jaredpar — APPROVED** [review]

**Comment by 333fred** [comment]
> Please look for other instances where we perform similar kind check or a type check and confirm that we are not missing a check for function pointers.

Looked through these cases. I only found one assert in the local rewriter around ensuring that a compound assignment LHS that got to this point has a ref-returning method.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CompoundAssignmentOperator.cs [hunk 2]
```diff
@@ -632,6 +632,10 @@ private BoundExpression TransformCompoundAssignmentLHS(BoundExpression originalL
                     Debug.Assert(((BoundCall)originalLHS).Method.RefKind != RefKind.None);
                     break;
 
+                case BoundKind.FunctionPointerInvocation:
+                    Debug.Assert(((BoundFunctionPointerInvocation)originalLHS).FunctionPointer.Signature.RefKind != RefKind.None);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CompoundAssignmentOperator.cs:636]: >Debug.Assert(((BoundFunctionPointerInvocation)originalLHS).FunctionPointer.Signature.RefKind != RefKind.None); [](start = 20, length = 110)

Do we have a test covering this code path?

Comment (333fred) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CompoundAssignmentOperator.cs:636]: Yep, the positive test I added hits this.

**Pull request merged**